File: sdks/go/reproit/distributed_fuzz.go
package reproit

import (
	"bytes"
	"context"
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"regexp"
	"time"
)

const (
	FuzzContextHTTPHeader   = "ReproIt-Fuzz-Context"
	FuzzParentHTTPHeader    = "ReproIt-Parent-Operation"
	FuzzContextQueueField   = "reproit.fuzz.context"
	FuzzParentQueueField    = "reproit.parent.operation"
	maximumFuzzContextBytes = 4_096
)

var (
	fuzzCampaignIDPattern = regexp.MustCompile(
		`^fc_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`,
	)
	fuzzCaseIDPattern = regexp.MustCompile(
		`^case_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`,
	)
	fuzzProjectIDPattern = regexp.MustCompile(
		`^prj_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`,
	)
	fuzzServiceIDPattern = regexp.MustCompile(
		`^svc_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`,
	)
	fuzzOperationIDPattern = regexp.MustCompile(
		`^op_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`,
	)
)

var ErrFuzzContext = errors.New("Repro It rejected the fuzz context.")

type signedFuzzContext struct {
	CampaignID string `json:"campaign_id"`
	CaseID     string `json:"case_id"`
	ExpiresAt  string `json:"expires_at"`
	Format     string `json:"format"`
	ProjectID  string `json:"project_id"`
	ServiceID  string `json:"service_id"`
	Signature  string `json:"signature"`
}

type FuzzContextValidator struct {
	Clock           func() time.Time
	ProjectID       string
	VerificationKey []byte
}

type FuzzCampaignContext struct {
	context         signedFuzzContext
	contextDigest   string
	encoded         string
	now             string
	parentOperation string
	verificationKey string
}

type fuzzContextKey struct{}

func (validator FuzzContextValidator) Validate(encoded string) (*FuzzCampaignContext, error) {
	if len(encoded) == 0 || len(encoded) > 5_462 || len(validator.VerificationKey) != 32 ||
		validator.Clock == nil {
		return nil, ErrFuzzContext
	}
	decoded, err := base64.RawURLEncoding.DecodeString(encoded)
	if err != nil || len(decoded) > maximumFuzzContextBytes {
		return nil, ErrFuzzContext
	}
	var value signedFuzzContext
	decoder := json.NewDecoder(bytes.NewReader(decoded))
	decoder.DisallowUnknownFields()
	if decoder.Decode(&value) != nil || decoder.Decode(&struct{}{}) != io.EOF {
		return nil, ErrFuzzContext
	}
	signed, err := CanonicalBytes(fuzzContextValue(value))
	if err != nil || string(signed) != string(decoded) || !validFuzzContext(value) ||
		value.ProjectID != validator.ProjectID {
		return nil, ErrFuzzContext
	}
	expiresAt, err := time.Parse("2006-01-02T15:04:05.000Z", value.ExpiresAt)
	if err != nil {
		return nil, ErrFuzzContext
	}
	now := validator.Clock().UTC()
	if !now.Before(expiresAt) {
		return nil, ErrFuzzContext
	}
	signature, err := base64.RawURLEncoding.DecodeString(value.Signature)
	if err != nil || len(signature) != ed25519.SignatureSize {
		return nil, ErrFuzzContext
	}
	unsigned := value
	unsigned.Signature = ""
	message, err := CanonicalBytes(fuzzContextValue(unsigned))
	if err != nil || !ed25519.Verify(validator.VerificationKey, message, signature) {
		return nil, ErrFuzzContext
	}
	digest := sha256.Sum256(signed)
	return &FuzzCampaignContext{
		context:         value,
		contextDigest:   "sha256:" + hex.EncodeToString(digest[:]),
		encoded:         encoded,
		now:             now.Format("2006-01-02T15:04:05.000Z"),
		verificationKey: base64.RawURLEncoding.EncodeToString(validator.VerificationKey),
	}, nil
}

func ExtractHTTPFuzzContext(
	parent context.Context,
	headers http.Header,
	validator FuzzContextValidator,
) (context.Context, error) {
	if parent == nil || headers == nil {
		return nil, ErrFuzzContext
	}
	encoded := headers.Get(FuzzContextHTTPHeader)
	parentID := headers.Get(FuzzParentHTTPHeader)
	if encoded == "" {
		if parentID != "" {
			return nil, ErrFuzzContext
		}
		return parent, nil
	}
	fuzzContext, err := validator.Validate(encoded)
	if err != nil || (parentID != "" && !fuzzOperationIDPattern.MatchString(parentID)) {
		return nil, ErrFuzzContext
	}
	fuzzContext.parentOperation = parentID
	return context.WithValue(parent, fuzzContextKey{}, fuzzContext), nil
}

func ExtractQueueFuzzContext(
	parent context.Context,
	metadata map[string]string,
	validator FuzzContextValidator,
) (context.Context, error) {
	headers := make(http.Header, 2)
	if encoded := metadata[FuzzContextQueueField]; encoded != "" {
		headers.Set(FuzzContextHTTPHeader, encoded)
	}
	if parentID := metadata[FuzzParentQueueField]; parentID != "" {
		headers.Set(FuzzParentHTTPHeader, parentID)
	}
	return ExtractHTTPFuzzContext(parent, headers, validator)
}

func PropagateQueueFuzzContext(ctx context.Context, metadata map[string]string) {
	fuzzContext, ok := fuzzContextFromContext(ctx)
	if !ok || metadata == nil {
		return
	}
	metadata[FuzzContextQueueField] = fuzzContext.encoded
	if fuzzContext.parentOperation != "" {
		metadata[FuzzParentQueueField] = fuzzContext.parentOperation
	}
}

func fuzzContextFromContext(ctx context.Context) (*FuzzCampaignContext, bool) {
	if ctx == nil || ctx.Err() != nil {
		return nil, false
	}
	value, ok := ctx.Value(fuzzContextKey{}).(*FuzzCampaignContext)
	return value, ok && value != nil
}

func (value *FuzzCampaignContext) beginIdentity() map[string]any {
	return map[string]any{
		"campaign_id":    value.context.CampaignID,
		"case_id":        value.context.CaseID,
		"context_digest": value.contextDigest,
	}
}

func (value *FuzzCampaignContext) nativeInput() *sdkEngineFuzzContextInput {
	return &sdkEngineFuzzContextInput{
		Encoded:         value.encoded,
		Now:             value.now,
		ProjectID:       value.context.ProjectID,
		ServiceID:       value.context.ServiceID,
		VerificationKey: value.verificationKey,
	}
}

func (value *FuzzCampaignContext) withParent(operationID string) *FuzzCampaignContext {
	child := *value
	child.parentOperation = operationID
	return &child
}

func validFuzzContext(value signedFuzzContext) bool {
	return value.Format == "reproit.fuzz-context.v1" &&
		fuzzCampaignIDPattern.MatchString(value.CampaignID) &&
		fuzzCaseIDPattern.MatchString(value.CaseID) &&
		fuzzProjectIDPattern.MatchString(value.ProjectID) &&
		fuzzServiceIDPattern.MatchString(value.ServiceID) && len(value.Signature) == 86
}

func fuzzContextValue(value signedFuzzContext) map[string]any {
	return map[string]any{
		"campaign_id": value.CampaignID,
		"case_id":     value.CaseID,
		"expires_at":  value.ExpiresAt,
		"format":      value.Format,
		"project_id":  value.ProjectID,
		"service_id":  value.ServiceID,
		"signature":   value.Signature,
	}
}

File: sdks/dotnet/ReproIt.Sdk/DistributedFuzz.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace ReproIt.Sdk;

/// <summary>Reports a rejected distributed fuzz context.</summary>
public sealed class FuzzContextException() : Exception(
    "Repro It rejected the fuzz context.");

/// <summary>Validates bounded context before native signature verification.</summary>
public sealed record FuzzContextValidator(
    string ProjectId,
    string VerificationKey)
{
    /// <summary>Validates one opaque context for an operation start.</summary>
    public FuzzCampaignContext Validate(string encoded, string now)
    {
        try
        {
            byte[] raw = DecodeBase64Url(encoded, 4_096);
            byte[] key = DecodeBase64Url(VerificationKey, 32);
            JsonObject value = JsonNode.Parse(raw) as JsonObject ?? throw Error();
            if (!CanonicalJson.Bytes(value).AsSpan().SequenceEqual(raw) ||
                value.Count != 7 || key.Length != 32)
            {
                throw Error();
            }
            string campaignId = Required(value, "campaign_id");
            string caseId = Required(value, "case_id");
            string expiresAt = Required(value, "expires_at");
            string format = Required(value, "format");
            string projectId = Required(value, "project_id");
            string serviceId = Required(value, "service_id");
            byte[] signature = DecodeBase64Url(Required(value, "signature"), 64);
            if (format != "reproit.fuzz-context.v1" ||
                !DistributedFuzz.CampaignIdPattern.IsMatch(campaignId) ||
                !DistributedFuzz.CaseIdPattern.IsMatch(caseId) ||
                !DistributedFuzz.ProjectIdPattern.IsMatch(projectId) ||
                !DistributedFuzz.ServiceIdPattern.IsMatch(serviceId) ||
                projectId != ProjectId || signature.Length != 64 ||
                !TryTimestamp(now, out DateTimeOffset current) ||
                !TryTimestamp(expiresAt, out DateTimeOffset expires) || current >= expires)
            {
                throw Error();
            }
            string digest = "sha256:" + Convert.ToHexStringLower(SHA256.HashData(raw));
            return new FuzzCampaignContext(
                campaignId,
                caseId,
                digest,
                encoded,
                now,
                null,
                projectId,
                serviceId,
                VerificationKey);
        }
        catch (Exception error) when (error is not FuzzContextException)
        {
            throw Error();
        }
    }

    private static bool TryTimestamp(string value, out DateTimeOffset timestamp) =>
        DateTimeOffset.TryParseExact(
            value,
            "yyyy-MM-dd'T'HH:mm:ss.fff'Z'",
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
            out timestamp);

    private static string Required(JsonObject value, string name) =>
        value[name]?.GetValue<string>() is string selected && selected.Length > 0
            ? selected
            : throw Error();

    private static byte[] DecodeBase64Url(string value, int maximumBytes)
    {
        if (string.IsNullOrEmpty(value) || value.Length > 5_462 ||
            !DistributedFuzz.Base64UrlPattern.IsMatch(value))
        {
            throw Error();
        }
        string padded = value.Replace('-', '+').Replace('_', '/');
        padded += new string('=', (4 - padded.Length % 4) % 4);
        byte[] decoded = Convert.FromBase64String(padded);
        return decoded.Length <= maximumBytes ? decoded : throw Error();
    }

    private static FuzzContextException Error() => new();
}

/// <summary>Contains one validated campaign and case context.</summary>
public sealed class FuzzCampaignContext
{
    internal FuzzCampaignContext(
        string campaignId,
        string caseId,
        string contextDigest,
        string encoded,
        string now,
        string? parentOperationId,
        string projectId,
        string serviceId,
        string verificationKey)
    {
        CampaignId = campaignId;
        CaseId = caseId;
        ContextDigest = contextDigest;
        Encoded = encoded;
        Now = now;
        ParentOperationId = parentOperationId;
        ProjectId = projectId;
        ServiceId = serviceId;
        VerificationKey = verificationKey;
    }

    /// <summary>Gets the campaign identity.</summary>
    public string CampaignId { get; }
    /// <summary>Gets the case identity.</summary>
    public string CaseId { get; }
    /// <summary>Gets the canonical signed-context digest.</summary>
    public string ContextDigest { get; }
    /// <summary>Gets the opaque signed context.</summary>
    public string Encoded { get; }
    /// <summary>Gets the causal parent operation.</summary>
    public string? ParentOperationId { get; }
    /// <summary>Gets the campaign project identity.</summary>
    public string ProjectId { get; }
    /// <summary>Gets the campaign root-service identity.</summary>
    public string ServiceId { get; }
    internal string Now { get; }
    internal string VerificationKey { get; }

    /// <summary>Activates the context for one synchronous or asynchronous flow.</summary>
    public IDisposable Activate() => DistributedFuzz.Activate(this);

    internal FuzzCampaignContext WithParent(string operationId)
    {
        if (!DistributedFuzz.OperationIdPattern.IsMatch(operationId))
        {
            throw new FuzzContextException();
        }
        return new FuzzCampaignContext(
            CampaignId,
            CaseId,
            ContextDigest,
            Encoded,
            Now,
            operationId,
            ProjectId,
            ServiceId,
            VerificationKey);
    }

    internal JsonObject BeginIdentity() => new()
    {
        ["campaign_id"] = CampaignId,
        ["case_id"] = CaseId,
        ["context_digest"] = ContextDigest,
    };

    internal JsonObject NativeInput() => new()
    {
        ["encoded"] = Encoded,
        ["now"] = Now,
        ["project_id"] = ProjectId,
        ["service_id"] = ServiceId,
        ["verification_key"] = VerificationKey,
    };
}

/// <summary>Extracts and propagates distributed fuzz context.</summary>
public static partial class DistributedFuzz
{
    /// <summary>The inbound and outbound HTTP context header.</summary>
    public const string ContextHttpHeader = "ReproIt-Fuzz-Context";
    /// <summary>The inbound and outbound HTTP parent header.</summary>
    public const string ParentHttpHeader = "ReproIt-Parent-Operation";
    /// <summary>The delivered-work context metadata field.</summary>
    public const string ContextQueueMetadata = "reproit.fuzz.context";
    /// <summary>The delivered-work parent metadata field.</summary>
    public const string ParentQueueMetadata = "reproit.parent.operation";

    private static readonly AsyncLocal<FuzzCampaignContext?> Current = new();

    [GeneratedRegex(
        "^fc_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")]
    internal static partial Regex CampaignIdRegex();
    [GeneratedRegex(
        "^case_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")]
    internal static partial Regex CaseIdRegex();
    [GeneratedRegex(
        "^prj_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")]
    internal static partial Regex ProjectIdRegex();
    [GeneratedRegex(
        "^svc_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")]
    internal static partial Regex ServiceIdRegex();
    [GeneratedRegex(
        "^op_[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")]
    internal static partial Regex OperationIdRegex();
    [GeneratedRegex("^[A-Za-z0-9_-]+$")]
    internal static partial Regex Base64UrlRegex();

    internal static Regex CampaignIdPattern => CampaignIdRegex();
    internal static Regex CaseIdPattern => CaseIdRegex();
    internal static Regex ProjectIdPattern => ProjectIdRegex();
    internal static Regex ServiceIdPattern => ServiceIdRegex();
    internal static Regex OperationIdPattern => OperationIdRegex();
    internal static Regex Base64UrlPattern => Base64UrlRegex();

    /// <summary>Validates inbound HTTP metadata.</summary>
    public static FuzzCampaignContext? ExtractHttp(
        IReadOnlyDictionary<string, string> headers,
        FuzzContextValidator validator,
        string now)
    {
        string? encoded = Header(headers, ContextHttpHeader);
        string? parent = Header(headers, ParentHttpHeader);
        if (encoded is null)
        {
            return parent is null ? null : throw new FuzzContextException();
        }
        FuzzCampaignContext context = validator.Validate(encoded, now);
        return parent is null ? context : context.WithParent(parent);
    }

    /// <summary>Validates inbound delivered-work metadata.</summary>
    public static FuzzCampaignContext? ExtractQueue(
        IReadOnlyDictionary<string, string> metadata,
        FuzzContextValidator validator,
        string now)
    {
        Dictionary<string, string> headers = [];
        if (metadata.TryGetValue(ContextQueueMetadata, out string? encoded))
        {
            headers[ContextHttpHeader] = encoded;
        }
        if (metadata.TryGetValue(ParentQueueMetadata, out string? parent))
        {
            headers[ParentHttpHeader] = parent;
        }
        return ExtractHttp(headers, validator, now);
    }

    /// <summary>Adds active context to outbound delivered-work metadata.</summary>
    public static void PropagateQueue(IDictionary<string, string> metadata)
    {
        FuzzCampaignContext? context = Active();
        if (context is null)
        {
            return;
        }
        metadata[ContextQueueMetadata] = context.Encoded;
        if (context.ParentOperationId is not null)
        {
            metadata[ParentQueueMetadata] = context.ParentOperationId;
        }
    }

    internal static FuzzCampaignContext? Active() =>
        AutomaticOperationContext.ActiveOperation()?.FuzzContext ?? Current.Value;

    internal static IDisposable Activate(FuzzCampaignContext context)
    {
        FuzzCampaignContext? parent = Current.Value;
        Current.Value = context;
        return new Activation(context, parent);
    }

    private static string? Header(
        IReadOnlyDictionary<string, string> headers,
        string expected)
    {
        List<string> values = headers
            .Where(pair => string.Equals(pair.Key, expected, StringComparison.OrdinalIgnoreCase))
            .Select(pair => pair.Value)
            .ToList();
        return values.Count switch
        {
            0 => null,
            1 when !string.IsNullOrEmpty(values[0]) => values[0],
            _ => throw new FuzzContextException(),
        };
    }

    private sealed class Activation(
        FuzzCampaignContext context,
        FuzzCampaignContext? parent) : IDisposable
    {
        private bool disposed;

        public void Dispose()
        {
            if (disposed)
            {
                return;
            }
            disposed = true;
            if (ReferenceEquals(Current.Value, context))
            {
                Current.Value = parent;
            }
        }
    }
}
